kernel/packages/unity-interface/BrowserInterface.ts

```diff
@@ -621,6 +631,41 @@ export class BrowserInterface {
   private arrayCleanup<T>(array: T[] | null | undefined): T[] | undefined {
     return !array || array.length === 0 ? undefined : array
   }
+
+  private async JumpToAnotherRealm(store: Store<RootState>, nextRealmIndexToTry: number, jumpInData: JumpInPayload) {
+    const realmString = jumpInData.realm.serverName + '-' + jumpInData.realm.layer
+
+    if (nextRealmIndexToTry < jumpInData.nextMostPopulatedRealms.length) {
+      notifyStatusThroughChat(`Couldn't find realm ${realmString}. Trying to connect to the next more populated one...`)
+      await this.TryToJumpToTheNextMostPopulatedRealm(store, nextRealmIndexToTry, jumpInData)
+    } else {
+      notifyStatusThroughChat(`Couldn't find realm ${realmString}. You'll stay in your current realm.`)
+      this.GoTo(jumpInData.gridPosition)
+    }
+  }
+
+  private async TryToJumpToTheNextMostPopulatedRealm(store: Store<RootState>, realmIndex: number, jumpInData: JumpInPayload) {
```

Comment:
I see that this method doesn't do anything that we couldn't do inside Unity.

We can't just expose the `JumpIn` method (with failure callbacks, etc.) and do all the handling from the renderer? Otherwise, this approach would violate our current direction of moving responsibilities to Unity.

Reply:
➕

Reply:
Hi guys, I would like to discuss this suggested approach before start any refactor on this, because I think that in this specific case it is not the best idea to move these responsibilities to Unity.

This `TryToJumpToTheNextMostPopulatedRealm` function is being called by the `JumpIn` function each time a connection with a realm fails and, if this happens, this function is doing a recursive call to the `JumpIn` function again to make a new try with the next best realm. So the current flow is:

### Current approach
1. **[Unity] -> [Kernel]**: Unity calls `JumpIn` function in Kernel with the corresponding realms information.
2. **[Kernel]** Kernel try to connect to the first candidate realm.
2.1. **[Kernel]** If the connection is successfully stablished, it does the jump to the corresponding scene/realm.
2.2. **[Kernel]** If the connection fails, it try with the next candidate realm (we go to the point 2).

_(Only 1 connection from Unity to Kernel, and Kernel is who have the responsibility of manage the connections)_

With the approach of moving the connection errors handling to Unity (@BrianAmadori correct me if I'm wrong), the flow would be the next:

### Suggested approach
1. **[Unity] -> [Kernel]**: Unity calls `JumpIn` function in Kernel with the first candidate realm information.
2. **[Kernel]** Kernel try to connect to this realm.
2.1. **[Kernel]** If the connection is successfully stablished, it does the jump to this scene/realm.
2.2. **[Kernel] -> [Unity]** If the connection fails, Kernel will send a message to Unity to inform about the connection error.
3. **[Unity]**: Unity receives an error message from Kernel, then it calls again the `JumpIn` function in Kernel with the next candidate realm information (we go to the point 2).

_(There could be a lot of connections between Unity and Kernel. Kernel would only have the responsibility of connect with a specifi realm and Unity would have the responsibility of manage the "retries" system)_

For me, both approachs are valid, but I would personally prefer to avoid so much back-and-forth between Unity and Kernel, and give all the responsibility of "connection retries" to Kernel, which makes the connections with the catalysts.

Anyway, please let review both flows and let me know what do you think. If we finally decide to go with the second approach, we can do the refactor, no problem! :)

Reply:
I think your argument makes sense from a mechanical complexity standpoint. The direction of moving the Kernel responsibilities over to Unity is reasoned from a maintenance costs standpoint.

As we argued many times, we should try to move most of the business logic maintenance responsibilities over to Unity, leaving to Kernel the low-level queries and features that can't be done from Unity. This is NOT done to reduce the roundtrips between Kernel and Unity, but to make maintenance easier (aka less costly).

If the approach of this PR is put on place, then the maintenance of the Explore feature will suddenly get harder because we increase the maintenance cost. I argue that the cost gets higher the more code we have on both components for the same feature. When you need to maintain features that have a Kernel and Unity side running at the same time, TypeScript understanding is needed, kernel<>renderer interoperability knowledge is needed, having to use make watch to test, double PRs, etc.

To rephrase and make the argument as clear as possible: It's true that from a mechanical standpoint we get a lot of roundtrips, but here I think we can trade off the roundtrips to cheapen the maintenance cost--we don't have any performance bottlenecks anyway. 

Also, this would enable the possibility of handling custom jumps in retry logic for other features entirely from Unity too.

Reply:
Ok! Let follow this approach then. Thank you Brian!

Reply:
@BrianAmadori I have already applied the agreed refactor: Now, the Kernel side remained as it was but only adding the 2 needed callbacks (`ConnectionToRealmSuccess` and `ConnectionToRealmFailed`) for informing Unity about the result of the JumpIn. And all the logic related to the realms retries system has been moved to the Unity side (https://github.com/decentraland/unity-renderer/pull/781).

Could you please take a look to both PRs? Thanks!

Reply:
Done!